File: problems/0701-0800/0787_minimum-cost-with-limited-intermediates/solution_dijkstra_budget.js
/**
 * @param {number} nodeCount
 * @param {number[][]} links
 * @param {number} source
 * @param {number} target
 * @param {number} maxIntermediates
 * @return {number}
 */
var minimumLimitedRouteCost = function (nodeCount, links, source, target, maxIntermediates) {
    const graph = Array.from({ length: nodeCount }, () => []);
    for (const [f, t, weight] of links) {
        graph[f].push([t, weight]);
    }

    // Min-heap of [cost, node, links taken]
    const heap = [];
    const push = (item) => {
        heap.push(item);
        let i = heap.length - 1;
        while (i > 0) {
            const p = (i - 1) >> 1;
            if (heap[p][0] <= heap[i][0]) break;
            [heap[p], heap[i]] = [heap[i], heap[p]];
            i = p;
        }
    };
    const pop = () => {
        const top = heap[0];
        const last = heap.pop();
        if (heap.length > 0) {
            heap[0] = last;
            let i = 0;
            for (;;) {
                const l = 2 * i + 1;
                const r = l + 1;
                let m = i;
                if (l < heap.length && heap[l][0] < heap[m][0]) m = l;
                if (r < heap.length && heap[r][0] < heap[m][0]) m = r;
                if (m === i) break;
                [heap[m], heap[i]] = [heap[i], heap[m]];
                i = m;
            }
        }
        return top;
    };

    // State = (cost, node, links taken). Carrying the count in the state
    // is what enforces the limit: a state that already used its maxIntermediates+1 links
    // is never allowed to board another.
    push([0, source, 0]);
    const best = new Array(nodeCount).fill(Infinity);
    while (heap.length > 0) {
        const [cost, node, edges] = pop();
        // The heap pops in cost order, so the first target pop is final.
        if (node === target) {
            return cost;
        }
        // Dominance prune: a cheaper state that used no more links was
        // already expanded here, so this one cannot lead anywhere new.
        if (edges > best[node]) {
            continue;
        }
        best[node] = edges;
        if (edges < maxIntermediates + 1) {
            for (const [nxt, weight] of graph[node]) {
                push([cost + weight, nxt, edges + 1]);
            }
        }
    }
    return -1;
};

File: problems-adapt/0701-0800/0787_minimum-cost-with-limited-intermediates/solution_dijkstra_budget.go
import "container/heap"

type stopHeap [][3]int

func (h stopHeap) Len() int            { return len(h) }
func (h stopHeap) Less(i, j int) bool  { return h[i][0] < h[j][0] }
func (h stopHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *stopHeap) Push(x interface{}) { *h = append(*h, x.([3]int)) }
func (h *stopHeap) Pop() interface{} {
	old := *h
	nodeCount := len(old)
	item := old[nodeCount-1]
	*h = old[:nodeCount-1]
	return item
}

func minimumLimitedRouteCost(nodeCount int, links [][]int, source int, target int, maxIntermediates int) int {
	graph := make([][][2]int, nodeCount)
	for _, link := range links {
		graph[link[0]] = append(graph[link[0]], [2]int{link[1], link[2]})
	}
	const inf = int(^uint(0) >> 1)
	// State = (cost, node, links taken). Carrying the count in the state
	// is what enforces the limit: a state that already used its maxIntermediates+1 links
	// is never allowed to board another.
	h := &stopHeap{}
	heap.Push(h, [3]int{0, source, 0})
	best := make([]int, nodeCount)
	for i := range best {
		best[i] = inf
	}
	for h.Len() > 0 {
		top := heap.Pop(h).([3]int)
		cost, node, edges := top[0], top[1], top[2]
		// The heap pops in cost order, so the first target pop is final.
		if node == target {
			return cost
		}
		// Dominance prune: a cheaper state that used no more links was
		// already expanded here, so this one cannot lead anywhere new.
		if edges > best[node] {
			continue
		}
		best[node] = edges
		if edges < maxIntermediates+1 {
			for _, e := range graph[node] {
				heap.Push(h, [3]int{cost + e[1], e[0], edges + 1})
			}
		}
	}
	return -1
}
